receiver/statsdreceiver/transport/packet_server.go

```diff
@@ -0,0 +1,131 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package transport // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/statsdreceiver/transport"
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"strings"
+
+	"go.opentelemetry.io/collector/consumer"
+	"go.uber.org/multierr"
+)
+
+type packetServer struct {
+	packetConn net.PacketConn
+	transport  Transport
+	address    string
+}
+
+// Ensure that Server is implemented on Packet transport based servers.
+var _ (Server) = (*packetServer)(nil)
+
+// NewPacketServer creates a transport.Server using transports based on packets.
+func NewPacketServer(transport Transport, address string) (Server, error) {
+	var psrv packetServer
+	var err error
+
+	if !transport.IsPacketTransport() {
+		return nil, ErrUnsupportedPacketTransport
+	}
+	if transport.IsUnixTransport() {
+		err := os.Remove(address)
+		if err != nil && !errors.Is(err, os.ErrNotExist) {
+			return nil, fmt.Errorf("removing socket failed: %w", err)
+		}
+	}
+
+	psrv.transport = transport
+	psrv.address = address
+	psrv.packetConn, err = net.ListenPacket(transport.String(), address)
+	if err != nil {
+		return nil, fmt.Errorf("starting to listen %s socket: %w", psrv.transport, err)
+	}
+
+	return &psrv, nil
+}
+
+// Start the server ready to receive metrics.
+func (psrv *packetServer) ListenAndServe(
+	nextConsumer consumer.Metrics,
+	reporter Reporter,
+	transferChan chan<- Metric,
+) error {
+	if nextConsumer == nil || reporter == nil {
+		return errNilListenAndServeParameters
+	}
+
+	buf := make([]byte, 65527) // max size for udp packet body (assuming ipv6)
+	for {
+		n, addr, err := psrv.packetConn.ReadFrom(buf)
+
+		if _, ok := psrv.packetConn.(*net.UnixConn); ok && addr == nil {
+			addr = &net.UnixAddr{
+				Net:  "unixgram",
+				Name: "UDS",
+			}
+		}
+
+		if n > 0 {
+			bufCopy := make([]byte, n)
+			copy(bufCopy, buf)
+			psrv.handlePacket(bufCopy, addr, transferChan)
+		}
+		if err != nil {
+			var netErr net.Error
+			reporter.OnDebugf("%s Transport (%s) - ReadFrom error: %v", psrv.transport, psrv.packetConn.LocalAddr(), err)
+			if errors.As(err, &netErr) {
+				if netErr.Timeout() {
+					continue
+				}
+			}
+			return err
+		}
+	}
+}
+
+// Closes the server
+func (psrv *packetServer) Close() error {
+	var errs error
+
+	if psrv.transport.IsUnixTransport() {
+		err := os.Remove(psrv.address)
+		if err != nil && !errors.Is(err, os.ErrNotExist) {
+			errs = multierr.Append(errs, fmt.Errorf("removing socket failed: %w", err))
+		}
+	}
+
+	err := psrv.packetConn.Close()
+	if err != nil {
+		errs = multierr.Append(errs, err)
+	}
+
+	return errs
+}
+
+// This helper parses the buffer and split it line bye line to be parsed upstream.
+func (psrv *packetServer) handlePacket(
+	data []byte,
+	addr net.Addr,
+	transferChan chan<- Metric,
+) {
+	buf := bytes.NewBuffer(data)
+	for {
+		bytes, err := buf.ReadBytes((byte)('\n'))
+		if errors.Is(err, io.EOF) {
+			if len(bytes) == 0 {
```

Comment:
could this become:
if errors.Is(err, io.EOF) && len(bytes) == 0 {

Reply:
This has exactly the same structure as the receiver had before my refactor. If you do not go commit by commit you'll see a whole new file but if you do you see that I did not make any change on that part.